src/main/java/com/linecorp/armeria/common/logging/AbstractMessageLog.java

```diff
@@ -0,0 +1,200 @@
+        if (isDone()) {
+            return;
+        }
+
+        this.contentLength = contentLength;
+    }
+
+    @Override
+    public long contentLength() {
+        return contentLength;
+    }
+
+    @Override
+    public void attach(Object attachment) {
+        requireNonNull(attachment, "attachment");
+        if (isDone()) {
+            return;
+        }
+
+        if (attachments == null) {
+            attachments = new ArrayList<>(4);
+        }
+
+        attachments.add(attachment);
+    }
+
+    @Override
+    public List<Object> attachments() {
+        if (attachments != null) {
+            return attachments;
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
+    @Override
+    public <A> A attachment(Class<A> type) {
+        if (attachments == null) {
+            return null;
+        }
+
+        for (Object a : attachments) {
+            if (type.isInstance(a)) {
+                @SuppressWarnings("unchecked")
+                A cast = (A) a;
+                return cast;
+            }
+        }
+
+        return null;
+    }
+
+    @Override
+    public void end(Throwable cause) {
+        if (isDone()) {
+            return;
+        }
+
+        this.cause = cause;
+
+        if (attachments != null) {
+            final List<CompletableFuture<?>> dependencies = new ArrayList<>(attachments.size());
+            for (Object a : attachments) {
+                if (a instanceof CompletableFuture) {
+                    final CompletableFuture<?> f = (CompletableFuture<?>) a;
+                    if (!f.isDone()) {
+                        dependencies.add(f);
+                    }
+                }
+            }
+
+            final CompletableFuture<?> future;
+            switch (dependencies.size()) {
+                case 0:
+                    complete();
+                    return;
+                case 1:
+                    future = dependencies.get(0);
+                    break;
+                default:
+                    future = CompletableFuture.allOf(
+                            dependencies.toArray(new CompletableFuture<?>[dependencies.size()]));
+            }
+
+            future.whenComplete((unused1, unused2) -> complete());
+        } else {
+            complete();
+        }
+    }
+
+    private void complete() {
+        endTimeNanos = System.nanoTime();
+        complete(self());
+    }
+
+    @Override
+    public long endTimeNanos() {
+        return endTimeNanos;
+    }
+
+    @Override
+    public Throwable cause() {
+        return cause;
+    }
+
+    @SuppressWarnings("unchecked")
+    private T self() {
+        return (T) this;
+    }
+
+    @Override
+    public final String toString() {
+        final MoreObjects.ToStringHelper helper =
+                MoreObjects.toStringHelper("")
+                           .add("timeSpan",
+                                startTimeNanos + "+" + UnitFormatter.elapsed(startTimeNanos, endTimeNanos))
```

Comment:
> discuss: Also I'm not sure about auto-unit. personally I prefer fixed unit(maybe nano or micro)

I prefer auto-unit because different services have different range of processing time. For some services, sub-second precision has no meaning while other services may.